src/bartiq/transform.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2024 PsiQuantum, Corp.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import copy
+from typing import Any, Dict, Union
+
+import sympy
+from qref import SchemaV1
+
+def add_aggregated_resources(
+    aggregation_dict: Dict[str, Dict[str, Any]], qref_obj: Union[SchemaV1, dict]
+) -> Union[SchemaV1, dict]:
+    """
+    Add aggregated resources to qref program based on the aggregation dictionary.
+
+    Args:
+        aggregation_dict (Dict[str, Dict[str, Any]]): The aggregation dictionary.
+        qref_obj (Union[Dict[str, Any], Any]): The program to which the resources will be added.
+
+    Returns:
+        Union[SchemaV1, dict]: The program with aggregated resources.
+
+    Raises:
+        TypeError: If the input types are not valid qref object.
+    """
+    if not SchemaV1.model_validate(qref_obj):
+        raise TypeError("Must apply to a qref object.")
+
+    expanded_aggregation_dict = _expand_aggregation_dict(aggregation_dict)
+    qref_obj["program"] = _process_program(expanded_aggregation_dict, qref_obj["program"])
+    return qref_obj
+
+
+def _add_aggregated_resources_to_subroutine(
+    expanded_aggregation_dict: Dict[str, Dict[str, Any]], subroutine: Dict[str, Any]
+) -> Dict[str, Any]:
+    """
+    Add aggregated resources to the subroutine based on the aggregation dictionary.
+
+    Args:
+        expanded_aggregation_dict (Dict[str, Dict[str, Any]]): The expanded aggregation dictionary.
+        subroutine (Dict[str, Any]): The subroutine to which the resources will be added.
+
+    Returns:
+        Dict[str, Any]: The subroutine with aggregated resources.
+    """
+    if "resources" not in subroutine:
+        return subroutine
+
+    # Initialize a dictionary for aggregated resources
+    aggregated_resources = {res["name"]: copy.deepcopy(res) for res in subroutine["resources"]}
+    # Iterate over the original resources in the subroutine
+    for resource in subroutine["resources"]:
+        resource_name = resource["name"]
+        resource_type = resource["type"]
+
+        try:
+            resource_expr = sympy.sympify(resource["value"])
```

Comment:
[issue] Please use backend instead of bare sympy.

Reply:
I guess that's `backend.parse` ?
Or if you start using native `Bartiq` routines, then all the expressions should already be sympy expressions or ints/floats, so you might not even need this step at all.

Reply:
It's `backend.as_expression` 🙂